fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/ServerPlayerEntityMixin.java

```diff
@@ -90,6 +107,21 @@ private void onCopyFrom(ServerPlayerEntity oldPlayer, boolean alive, CallbackInf
 		ServerPlayerEvents.COPY_FROM.invoker().copyFromPlayer(oldPlayer, (ServerPlayerEntity) (Object) this, alive);
 	}
 
+	@Inject(method = "teleport(Lnet/minecraft/server/world/ServerWorld;DDDLjava/util/Set;FF)Z", at = @At("HEAD"), cancellable = true)
+	private void beforeTeleport(ServerWorld world, double destX, double destY, double destZ, Set<PositionFlag> flags, float yaw, float pitch, CallbackInfoReturnable<Boolean> cir) throws CommandSyntaxException {
+		ServerPlayerEntity player = (ServerPlayerEntity) (Object) this;
+		boolean allowed = ServerPlayerEvents.ALLOW_TELEPORT.invoker().allowTeleport(player, world, new Vec3d(destX, destY, destZ));
+		if (!allowed) {
+			throw new CommandSyntaxException(null, new LiteralMessage(""));
+		}
+	}
+
+	@Inject(method = "teleport(Lnet/minecraft/server/world/ServerWorld;DDDLjava/util/Set;FF)Z", at = @At("TAIL"), cancellable = true)
```

Comment:
There's no need for these to be `cancellable`. There's also a potential issue with the before event being called and the after event not being called if someone cancels the method. Not sure how worried we are about that, and if it'd be worth requiring loader 0.16 to get <https://github.com/LlamaLad7/MixinExtras/wiki/WrapMethod> instead.